### ISSUE TITLE
Cyclic Learning Rates

### ISSUE THREAD

**Issue by jimfleming** [fomorians/pyoneer#11]
# Feature Request

## Objective

Add support for 1-cycle cyclic learning rates from https://arxiv.org/abs/1708.07120 with support for learning rate tests and inversion for momentum.

## Motivation

CLR is a state of the art technique for obtaining learning rate ranges and annealing schedules.

### Use Cases

Any supervised learning problem would benefit from CLR.

## Design Proposal

Optimizers in TF Eager support callables for learning rate (and momentum, etc.) I propose a class pattern which deterministically exposes the value of the schedule when the class is called.

### Examples

```
learning_rate = CyclicSchedule(
    vmin=0.0,
    vmax=1.0,
    step_size=1000,
    half=True,
    global_step=global_step)
optimizer = tf.train.AdamOptimizer(learning_rate)
tf.summary.scalar('learning_rate', learning_rate())
```

### API

```
class CyclicSchedule:
    def __init__(self, vmin, vmax, step_size, half=False, global_step=None):
        self.vmin = vmin
        self.vmax = vmax
        self.step_size = step_size
        self.half = half
        if global_step is None:
            global_step = tf.train.get_or_create_global_step()
        self.global_step = global_step

    def __call__(self):
        return value
```

### Alternatives Considered

Without using callables the same behavior can be achieved using a variable which is updated during iteration. This places state on the learning rate and is therefore an additional variable which needs to be persisted in checkpoints. The proposal above would make learning_rate deterministic from the global_step.

```
learning_rate = tfe.Variable(0.0, trainable=False)
optimizer = tf.train.AdamOptimizer(learning_rate)
learning_rate.assign(cyclic_schedule(
    vmin=0.0,
    vmax=1.0,
    step_size=1000,
    half=True,
    global_step=global_step))
tf.summary.scalar('learning_rate', learning_rate)
```

The same behavior as proposed could be accomplished with an inner function however this isn't a common pattern in Eager:

```
def cyclic_schedule(vmin, vmax, step_size, half, global_step):
    def _inner():
        return value

learning_rate = cyclic_schedule(
    vmin=0.0,
    vmax=1.0,
    step_size=1000,
    half=True,
    global_step=global_step)
optimizer = tf.train.AdamOptimizer(learning_rate)
tf.summary.scalar('learning_rate', learning_rate())
```
